src/platform/silabs/tracing/SilabsPowerTracing.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#if defined(SL_RAIL_LIB_MULTIPROTOCOL_SUPPORT) && SL_RAIL_LIB_MULTIPROTOCOL_SUPPORT
+#include "cmsis_os2.h"
+#include <sl_power_manager.h>
+#else
+// Use stub headers for unit tests
+#include <platform/silabs/tracing/mocks/cmsis_os2_stub.h>        // nogncheck
+#include <platform/silabs/tracing/mocks/sl_power_manager_stub.h> // nogncheck
+#endif
+
+#include <lib/core/CHIPError.h>
+
+namespace chip {
+namespace Tracing {
+namespace Silabs {
+
+struct EnergyTrace
+{
+    uint32_t mEntryTime;
+    sl_power_manager_em_t mEnergyMode;
+};
+
+class SilabsPowerTracing
+{
+public:
+    SilabsPowerTracing();
+    ~SilabsPowerTracing();
+
+    static SilabsPowerTracing & Instance() { return sInstance; }
+
+    /** @brief Initialize the power tracing system
+     * Allocates energy trace storage, subscribes to power manager events, and starts the statistics timer.
+     * @return CHIP_NO_ERROR on success, CHIP_ERROR_NO_MEMORY if allocation fails, CHIP_ERROR_INTERNAL on timer errors
+     */
+    CHIP_ERROR Init();
+
+    /** @brief Output all recorded power manager energy mode traces
+     * Logs the collected energy traces showing timestamps and energy modes to the device layer.
+     * @return CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR OutputPowerManagerTraces();
+
+    /** @brief Check if the power tracing system is initialized
+     *  @return true if initialized, false otherwise
+     */
+    bool IsInitialized() const { return mInitialized; }
+
+    /** @brief Get the number of recorded energy traces
+     *  @return The count of energy traces currently in the buffer
+     */
+    size_t GetEnergyTraceCount() const { return mEnergyTraceCount; }
+
+    /** @brief Get a specific energy trace by index
+     *  @param index The index of the trace to retrieve
+     *  @return Pointer to the energy trace, or nullptr if index is out of bounds
+     */
+    const EnergyTrace * GetEnergyTrace(size_t index) const
+    {
+        if (index >= mEnergyTraceCount || mEnergyTraces == nullptr)
```

Comment:
For reference, checks happen in the order they are written.

That being said, I would check mInitialized here. If we deem there are possibilities where mInitialized is true and mEnergyTraces is false, we should check both.